components/eamxx/src/physics/mam/eamxx_mam_microphysics_process_interface.hpp

```diff
@@ -0,0 +1,369 @@
+#include <share/atm_process/ATMBufferManager.hpp>
+
+#include <ekat/ekat_parameter_list.hpp>
+#include <ekat/ekat_workspace.hpp>
+#include <ekat/logging/ekat_logger.hpp>
+#include <mam4xx/mam4.hpp>
+
+#include <string>
+
+#ifndef KOKKOS_ENABLE_CUDA
+#define protected public
+#define private public
+#endif
+
+namespace scream
+{
+
+// The process responsible for handling MAM4 aerosols. The AD stores exactly ONE
+// instance of this class in its list of subcomponents.
+class MAMMicrophysics final : public scream::AtmosphereProcess {
+  using PF = scream::PhysicsFunctions<DefaultDevice>;
+  using KT = ekat::KokkosTypes<DefaultDevice>;
+
+  // views for single- and multi-column data
+  using view_1d_int   = typename KT::template view_1d<int>;
+  using view_1d       = typename KT::template view_1d<Real>;
+  using view_2d       = typename KT::template view_2d<Real>;
+  using const_view_1d = typename KT::template view_1d<const Real>;
+  using const_view_2d = typename KT::template view_2d<const Real>;
+
+  // unmanaged views (for buffer and workspace manager)
+  using uview_1d = Unmanaged<typename KT::template view_1d<Real>>;
+  using uview_2d = Unmanaged<typename KT::template view_2d<Real>>;
+
+  // a quantity stored in a single vertical column with a single index
+  using ColumnView = mam4::ColumnView;
+
+  // a thread team dispatched to a single vertical column
+  using ThreadTeam = mam4::ThreadTeam;
+
+  // a logger for this process
+  using Logger = ekat::logger::Logger<ekat::logger::LogNoFile,
+                                      ekat::logger::LogRootRank>;
+
+public:
+
+  // Constructor
+  MAMMicrophysics(const ekat::Comm& comm, const ekat::ParameterList& params);
+
+protected:
+
+  // --------------------------------------------------------------------------
+  // AtmosphereProcess overrides (see share/atm_process/atmosphere_process.hpp)
+  // --------------------------------------------------------------------------
+
+  // process metadata
+  AtmosphereProcessType type() const override;
+  std::string name() const override;
+
+  // grid
+  void set_grids(const std::shared_ptr<const GridsManager> grids_manager) override;
+
+  // management of common atm process memory
+  size_t requested_buffer_size_in_bytes() const override;
+  void init_buffers(const ATMBufferManager &buffer_manager) override;
+
+  // process behavior
+  void initialize_impl(const RunType run_type) override;
+  void run_impl(const double dt) override;
+  void finalize_impl() override;
+
+  // performs some checks on the tracers group
+  void set_computed_group_impl(const FieldGroup& group) override;
+
+private:
+
+  Logger logger;
+
+  // number of horizontal columns and vertical levels
+  int ncol_, nlev_;
+
+  // Atmosphere processes often have a pre-processing step that constructs
+  // required variables from the set of fields stored in the field manager.
+  // This functor implements this step, which is called during run_impl.
+  struct Preprocess {
```

Comment:
I know we do it in P3 and SHOC too, so this is fine, but the pre/post process structs (and their impl) are not really well fit in this header. Perhaps a pimpl would be appropriate here.

Reply:
Yeah, that might be best in the end. Do you mind if we wait till things are further along before we do this sort of stuff? The pimpl pattern would require us to move more code out of the header (which is good, but I prefer getting things to work "at all" before cleaning up).